examples/aggregation.py

```diff
@@ -0,0 +1,184 @@
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""Aggregation of probe results over bdb files and model numbers.
+
+A single probe is a typed result of some computation over a BayesDB,
+whose variation with model count, analysis iteration, or other setup
+one is interested in studying.
+
+A probe function is a function that accepts a bdb and computes a list
+of probe results (more than one so that they can share intermediate
+computations).
+
+A probe result is represented as a 3-tuple: the name of the probe, the
+type of the result (as a string), and the value of the result.
+
+Two result types are currently supported:
+- 'num' results are aggregated by maintaining a list of all of them.
+- 'bool' results are aggregated by maintaining the count of True and
+  False results seen.
+
+"""
+
+import contextlib
+import random
+import time
+
+import bayeslite
+from bayeslite.core import bayesdb_get_generator
+from bayeslite.metamodels.crosscat import CrosscatMetamodel
+from crosscat.LocalEngine import LocalEngine as CrosscatLocalEngine
+
+then = time.time()
+def log(msg):
+    print "At %3.2fs" % (time.time() - then), msg
+
+def analyze_fileset(files, generator, probes, model_schedule=None,
+                    n_replications=None, seed=0):
+    """Aggregate all the probes over all the given bdb files.
+
+    Do seed and model count variation within each file by varying
+    model sets to which probes are applied.
+
+    `files` is a list of filenames of saved BayesDBs to aggregate over.
+
+    `generator` is a string naming the generator the probes read.
+
+    `probes` is a list of probe functions.
+
+    `model_schedule` is a list of numbers of models to probe.
+
+    `n_replications` is the number of replications to do.
+
+    `seed` is the initial entropy for any randomness embedded in
+    BayesDB queries performed by the probe functions.  Default: 0.
+
+    Each saved bdb must have the named generator, and that generator
+    must have at least as many models as `n_replications *
+    max(model_schedule)`.
+
+    Probe functions are expected to utilize all available models to
+    compute their results; application to different sets of models is
+    accomplished by temporarily dropping unwanted models.
+
+    Returns a dict mapping conditions to aggregated probe results.
+    Each condition is a 3-tuple: Name of bdb file, number of models
+    probed, name of probe.  The results are aggregated over
+    replications: different sets of the desired number of models
+    within each file.
+
+    """
+    # TODO Expect the number of models in the file to be at least
+    # model_skip * n_replications; excess models are wasted.
+    # TODO Default the model schedule and n_replications based on the
+    # square root of the number of models in the first file
```

Comment:
Right.  Fixed in 6e3e667d110314f96a837cf96137b52b1f84a7b8 .

I should take this exchange as a lesson about reviewing in that style.